command/agent/acl_endpoint.go

```diff
@@ -676,3 +676,150 @@ func (s *HTTPServer) aclAuthMethodUpsertRequest(
+// API and handles read via both the ID, updates, and deletions.
+func (s *HTTPServer) ACLBindingRuleSpecificRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+
+	// Grab the suffix of the request, so we can further understand it.
+	reqSuffix := strings.TrimPrefix(req.URL.Path, "/v1/acl/binding-rule/")
+
+	// Ensure the binding rule ID is not an empty string which is possible if
+	// the caller requested "/v1/acl/role/binding-rule/".
+	if reqSuffix == "" {
+		return nil, CodedError(http.StatusBadRequest, "missing ACL binding rule ID")
+	}
+
+	// Identify the HTTP method which indicates which downstream function
+	// should be called.
+	switch req.Method {
+	case http.MethodGet:
+		return s.aclBindingRuleGetRequest(resp, req, reqSuffix)
+	case http.MethodDelete:
+		return s.aclBindingRuleDeleteRequest(resp, req, reqSuffix)
+	case http.MethodPost, http.MethodPut:
+		return s.aclBindingRuleUpsertRequest(resp, req, reqSuffix)
+	default:
+		return nil, CodedError(http.StatusMethodNotAllowed, ErrInvalidMethod)
+	}
+}
+
+func (s *HTTPServer) aclBindingRuleGetRequest(
+	resp http.ResponseWriter, req *http.Request, ruleID string) (interface{}, error) {
+
+	args := structs.ACLBindingRuleRequest{
+		ACLBindingRuleID: ruleID,
+	}
+	if s.parse(resp, req, &args.Region, &args.QueryOptions) {
+		return nil, nil
+	}
+
+	var reply structs.ACLBindingRuleResponse
+	if err := s.agent.RPC(structs.ACLGetBindingRuleRPCMethod, &args, &reply); err != nil {
+		return nil, err
+	}
+	setMeta(resp, &reply.QueryMeta)
+
+	if reply.ACLBindingRule == nil {
+		return nil, CodedError(http.StatusNotFound, "ACL binding rule not found")
+	}
+	return reply.ACLBindingRule, nil
+}
+
+func (s *HTTPServer) aclBindingRuleDeleteRequest(
+	resp http.ResponseWriter, req *http.Request, ruleID string) (interface{}, error) {
+
+	args := structs.ACLBindingRulesDeleteRequest{
+		ACLBindingRuleIDs: []string{ruleID},
+	}
+	s.parseWriteRequest(req, &args.WriteRequest)
+
+	var reply structs.ACLBindingRulesDeleteResponse
+	if err := s.agent.RPC(structs.ACLDeleteBindingRulesRPCMethod, &args, &reply); err != nil {
+		return nil, err
+	}
+	setIndex(resp, reply.Index)
+	return nil, nil
+}
+
+// aclBindingRuleUpsertRequest handles upserting an ACL binding rule to the
+// Nomad servers. It can handle both new creations, and updates to existing
+// rules.
+func (s *HTTPServer) aclBindingRuleUpsertRequest(
+	resp http.ResponseWriter, req *http.Request, ruleID string) (interface{}, error) {
+
+	// Decode the ACL binding rule.
+	var aclBindingRule structs.ACLBindingRule
+	if err := decodeBody(req, &aclBindingRule); err != nil {
+		return nil, CodedError(http.StatusBadRequest, err.Error())
+	}
+
+	// Ensure the request path ID matches the ACL binding rule ID that was
+	// decoded. Only perform this check on updates as a generic error on
+	// creation might be confusing to operators as there is no specific binding
+	// rule request path.
+	if ruleID != "" && ruleID != aclBindingRule.ID {
```

Comment:
Doesn't this condition imply that we allow users to set the ID on creation if they include it in the request body? The server is supposed to create it as a UUID.

Reply:
This function can be used by new creations or updates to existing binding rules. Therefore I feel it's worth checking that if the caller included an ID within the payload, that this is checked against the request path to ensure they match, otherwise we don't know which the correct one is. The RPC handler will check that the ID is found within state, indicating it is an update, before moving forward.

This comment did get me thinking about this function and I made a small modification within https://github.com/hashicorp/nomad/pull/15529/commits/4754d09f3d1f935e186d336a913f7ba40b4b78f3 which sets the aclBindingRule.ID to the request path ID, if the former has been left empty by the caller.